src/help.cpp

```diff
@@ -157,136 +115,216 @@ std::string help::get_note_colors()
-                    if( line_proc == "<DRAW_NOTE_COLORS>" ) {
-                        line_proc = get_note_colors();
-                    } else if( line_proc == "<HELP_DRAW_DIRECTIONS>" ) {
-                        line_proc = get_dir_grid();
-                    }
-                    size_t pos = line_proc.find( "<press_", 0, 7 );
-                    while( pos != std::string::npos ) {
-                        size_t pos2 = line_proc.find( ">", pos, 1 );
-
-                        std::string action = line_proc.substr( pos + 7, pos2 - pos - 7 );
-                        std::string replace = "<color_light_blue>" +
-                                              press_x( look_up_action( action ), "", "" ) + "</color>";
-
-                        if( replace.empty() ) {
-                            debugmsg( "Help json: Unknown action: %s", action );
-                        } else {
-                            line_proc = string_replace(
-                                            line_proc, "<press_" + std::move( action ) + ">", replace );
-                        }
-
-                        pos = line_proc.find( "<press_", pos2, 7 );
-                    }
-                    return line_proc;
-                } );
-
-                if( !i18n_help_texts.empty() ) {
-                    ui.on_screen_resize( nullptr );
-
-                    const auto get_w_help_border = [&]() {
-                        init_windows( ui );
-                        return w_help_border;
-                    };
-
-                    scrollable_text( get_w_help_border, _( "Help" ),
-                                     std::accumulate( i18n_help_texts.begin() + 1, i18n_help_texts.end(),
-                                                      i18n_help_texts.front(),
-                    []( std::string lhs, const std::string & rhs ) {
-                        return std::move( lhs ) + "\n\n" + rhs;
-                    } ) );
-
-                    ui.on_screen_resize( init_windows );
+void help_window::draw_category_option( const int &option, const help_category &category )
+{
+    std::string cat_name;
+    auto hotkey_it = hotkeys.find( option );
+    if( hotkey_it != hotkeys.end() ) {
+        cat_name = hotkey_it->second.short_description();
+        cat_name += ": ";
+    }
+    cat_name += category.name.translated();
+    if( data.read_categories.find( option ) != data.read_categories.end() ) {
+        ImGui::PushStyleColor( ImGuiCol_Text, c_light_gray );
+        ImGui::Selectable( remove_color_tags( cat_name ).c_str() );
+        ImGui::PopStyleColor();
+    } else {
+        ImGui::Selectable( remove_color_tags( cat_name ).c_str() );
+    }
+    if( ImGui::IsItemHovered() ) {
+        selected_option = option;
+    }
+}
+
+std::string help_window::seperator( int length, char c )
+{
+    std::string ret = "<color_light_blue>";
+    ret += std::string( length, c );
+    ret += "</color>";
+    return ret;
```

Comment:
Instead of allocating a new string with color tags and then immediately throwing it away, why not just do something like this?

```c++
ImGui::PushStyleColor( c_light_blue );
while ( c --> 0 ) {
    ImGui::TextUnformatted( c );
}
ImGui::PopStyleColor();
```

Reply:
I'm not sure it's worth emulating terminal style separators instead of just using the imgui ones, though.

Reply:
It’s also used for the border around the title, which I’m fine with. For extra points, use the gui width for the title, but make the border use the mono font as it currently does. You would want to measure the width of the title in pixels, then calculate the correct number of "=" will fit around it (since it won’t just be the same number of characters as the title).

You could also go nuts and add a large gui font. Give it +50% size (or maybe +25%, something that gives you a nice value when used with the default of 16px text size) and maybe a higher weight.

Reply:
I'm not having much luck with this, TextUnformatted() is adding a space after each call with SameLine (can't figure out what the SameLine args want) and I'm having to pass c as a string implicitly converted to a const char * bc TextUnformatted doesn't let me pass a character or 1 length string for whatever reason but then = isn't rendering while _ is?. (I'm just using regular single width = not the box drawing one)

> I'm not sure it's worth emulating terminal style separators instead of just using the imgui ones, though.

That's fair

> For extra points, use the gui width for the title, but make the border use the mono font as it currently does.

I saw that we can calc the pixel width of individual characters ye, I'll give it a go ^^

Reply:
Oh, for the character thing you could use `ImGui::Text("%c", c)` if you want. That avoids needing to allocate a two–byte string :) (Internally ImGui will use a fixed–sized buffer reserved for the purpose, iirc).